exoscale/loadbalancer.go

```diff
@@ -109,154 +97,193 @@ func (l *loadBalancer) EnsureLoadBalancer(ctx context.Context, _ string, service
-			return nil, err
-		}
-
-	default:
+	lbSpec, err := buildLoadBalancerFromAnnotations(service)
+	if err != nil {
 		return nil, err
 	}
 
-	_, err = l.fetchLoadBalancerService(lb, service)
-	switch err {
-	case nil:
-		if err := lb.UpdateService(ctx, lbDef.Services[0]); err != nil {
-			return nil, err
-		}
+	nlb, err := l.fetchLoadBalancer(ctx, service, zone)
+	if err != nil {
+		if errors.Is(err, errLoadBalancerNotFound) {
+			infof("creating new NLB %q", lbSpec.Name)
 
-	case errLoadBalancerServiceNotFound:
-		lbService, err := lb.AddService(ctx, lbDef.Services[0])
-		if err != nil {
-			return nil, err
-		}
+			nlb, err = l.p.client.CreateNetworkLoadBalancer(ctx, zone, lbSpec)
+			if err != nil {
+				return nil, err
+			}
+
+			if err := l.patchAnnotation(ctx, service, annotationLoadBalancerID, nlb.ID); err != nil {
+				return nil, fmt.Errorf("error patching annotations: %s", err)
+			}
 
-		if err := l.patchLoadBalancerServiceAnnotations(ctx, service, lbService); err != nil {
+			debugf("NLB %q created successfully (ID: %s)", nlb.Name, nlb.ID)
+		} else {
 			return nil, err
 		}
+	}
 
-	default:
+	if err = l.updateLoadBalancer(ctx, service); err != nil {
 		return nil, err
 	}
 
-	return &v1.LoadBalancerStatus{Ingress: []v1.LoadBalancerIngress{{IP: lb.IPAddress.String()}}}, nil
+	return &v1.LoadBalancerStatus{Ingress: []v1.LoadBalancerIngress{{IP: nlb.IPAddress.String()}}}, nil
 }
 
 // UpdateLoadBalancer updates hosts under the specified load balancer.
 // Implementations must treat the *v1.Service and *v1.Node
 // parameters as read-only and not modify them.
 // Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
-func (l *loadBalancer) UpdateLoadBalancer(ctx context.Context, _ string, service *v1.Service, nodes []*v1.Node) error {
+func (l *loadBalancer) UpdateLoadBalancer(ctx context.Context, _ string, service *v1.Service, _ []*v1.Node) error {
+	return l.updateLoadBalancer(ctx, service)
+}
+
+// EnsureLoadBalancerDeleted deletes the specified load balancer if it
+// exists, returning nil if the load balancer specified either didn't exist or
+// was successfully deleted.
+// This construction is useful because many cloud providers' load balancers
+// have multiple underlying components, meaning a Get could say that the LB
+// doesn't exist even if some part of it is still laying around.
+// Implementations must treat the *v1.Service parameter as read-only and not modify it.
+// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+func (l *loadBalancer) EnsureLoadBalancerDeleted(ctx context.Context, _ string, service *v1.Service) error {
 	zone, err := l.getLoadBalancerZone(service)
 	if err != nil {
 		return err
 	}
 
-	lbDef, err := buildLoadBalancerFromAnnotations(service)
+	nlb, err := l.fetchLoadBalancer(ctx, service, zone)
 	if err != nil {
+		if err == errLoadBalancerNotFound {
```

Comment:
If you want to stay consistent with your previous code, you can use
`errors.Is()` function.

Reply:
Good catch, fixed 👍